### PR TITLE
datatype: fixing dataloop MPIR_Typerep_pack

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Configuring ch4 with dataloop exposes a few bugs. This PR fixes them.

This PR fixes it.

[skip warnings]
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi

test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi

test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi

test:mpich/ch3/tcp
test:mpich/ch4/ofi

File: src/mpid/ch4/shm/posix/posix_am.h [hunk 1]
```diff
@@ -306,7 +306,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_do_am_isend(int grank,
 
     send_size = MPIDIG_am_send_async_get_data_sz_left(sreq);
     offset = MPIDIG_am_send_async_get_offset(sreq);
-    MPI_Aint *bytes_sent_out = MPIDIG_am_send_async_get_data_sz_left(sreq) ? &send_size : NULL;
+    MPI_Aint *bytes_sent_out = send_size ? &send_size : NULL;
```
Comment (raffenet) [inline src/mpid/ch4/shm/posix/posix_am.h:309]: this patch is good, but is anything every checking `bytes_sent_out`?

**Review by raffenet — APPROVED** [review]

File: src/mpid/ch4/shm/posix/posix_am.h (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/shm/posix/posix_am.h:309]: It's a pointer to `send_size`, which is updated by `MPIDI_POSIX_eager_send`. But it's a good point. I don't see why we need worry about passing `NULL` vs passing `&send_size` all the time. Let me change that.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/posix_am.h:309]: Oh, now I see it. Yes, probably can be improved 😄.

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Pull request merged**